### PR TITLE
Map Exchange to set up communication for MDLCSR

### PR DESCRIPTION
**Pull request by ywwu928**
# Description

## PR description

This PR implements map exchange to set up communication for MDLCSR.
Note that the main memory size and the heap size both needs to be increased to be able to run.
Changing the memory size from 8GB to 16GB requires rebuilding the container!

## Checklist

- [v] The additions follow the code standards in the [developer guide](pando-rt/docs/developer.md).
- [v] New or existing tests cover these changes.
- [v] The documentation is up to date with these changes.

### CODE REVIEW

File: include/pando-lib-galois/graphs/dist_local_csr.hpp [hunk 1]
```diff
@@ -693,6 +728,62 @@ class DistLocalCSR {
...
+    PANDO_CHECK(mirrorList.initialize());
+    auto createMirrors =
+        +[](galois::HostIndexedMap<pando::Vector<pando::Vector<ReadEdgeType>>> partEdges,
+            HostLocalStorage<pando::Array<std::uint64_t>> mirrorList,
+            HostLocalStorage<pando::Array<std::uint64_t>> V2PM, std::uint64_t i,
+            galois::WaitGroup::HandleType wgh) {
+          pando::Array<uint64_t> mirrors;
+
+          // Populating the mirror list in a set to avoid duplicates
+          std::unordered_set<uint64_t> mirrorMap;
```
Comment (AdityaAtulTewari) [inline include/pando-lib-galois/graphs/dist_local_csr.hpp:749]: This data structure cannot be used. instead use a hashtable to a boolean.

File: include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+  }
+
+  /** Sync **/
+  // TODO(Ying-Wei):
+  // write a sync function that reduces mirror values and then broadcasts master values
+  // return a bitmap of modified vertices
+  //
+  // template <typename Func>
+  // pando::Array<bool> sync(Func func, pando::Array<bool>) {
+  //}
```
Comment (AdityaAtulTewari) [inline include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp:262]: delete and save that for a separate pr.

File: include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+      PANDO_CHECK(
+          pando::executeOn(place, generateMetadata, *this, this->dlcsr, mirrorList, i, wgh));
+      numVertices += lift(mirrorList[i], size);
+    }
+    PANDO_CHECK(wg.wait());
+
+    PANDO_CHECK_RETURN(setupCommunication());
+
+    return pando::Status::Success;
+  }
```
Comment (AdityaAtulTewari) [inline include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp:404]: Is this PR supposed to include all of this work?

File: test/graphs/test_mirror_dist_local_csr.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                      std::make_tuple("/pando/graphs/rmat_571919_seed1_scale10_nV1024_nE10447.el",
+                                      1024)));
+
+INSTANTIATE_TEST_SUITE_P(
+    DISABLED_BigFiles, MirrorDLCSRInitEdgeList,
+    ::testing::Values(
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale11_nV2048_nE22601.el", 2048),
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale12_nV4096_nE48335.el", 4096),
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale13_nV8192_nE102016.el", 8192),
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale14_nV16384_nE213350.el", 16384)));
```
Comment (AdityaAtulTewari) [inline test/graphs/test_mirror_dist_local_csr.cpp:142]: when you delete DISABLED do these pass?

File: test/import/test_wmd_importer.cpp [hunk 5]
```diff
@@ -270,3 +271,135 @@ INSTANTIATE_TEST_SUITE_P(
...
+// std::make_tuple("/pando/graphs/rmat_571919_seed1_scale10_nV1024_nE10447.el",
+// 1024)));
+
+INSTANTIATE_TEST_SUITE_P(
+    DISABLED_BigFiles, MirrorDLCSRInitEdgeList,
+    ::testing::Values(
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale11_nV2048_nE22601.el", 2048),
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale12_nV4096_nE48335.el", 4096),
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale13_nV8192_nE102016.el", 8192),
+        std::make_tuple("/pando/graphs/rmat_571919_seed1_scale14_nV16384_nE213350.el", 16384)));
```
Comment (AdityaAtulTewari) [inline test/import/test_wmd_importer.cpp:405]: When you delete DISABLED does this pass?

**Review by AdityaAtulTewari — APPROVED** [review]

File: include/pando-lib-galois/graphs/dist_local_csr.hpp (hunk 1 above)
Comment (ywwu928) [reply include/pando-lib-galois/graphs/dist_local_csr.hpp:749]: @jeageun This is your code I think

File: include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp (hunk 3 above)
Comment (ywwu928) [reply include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp:404]: No, but the progress on MDLCSR construction PR is too slow so I have to base on the old version and work on it.
Do you want me to wait until the PR for MDLCSR construction to pass?

File: include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp (hunk 2 above)
Comment (ywwu928) [reply include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp]: done

File: test/graphs/test_mirror_dist_local_csr.cpp (hunk 4 above)
Comment (ywwu928) [reply test/graphs/test_mirror_dist_local_csr.cpp:142]: Yes, but it takes a bit longer. Do you want me to enable them?

File: test/import/test_wmd_importer.cpp (hunk 5 above)
Comment (ywwu928) [reply test/import/test_wmd_importer.cpp:405]: Yes, but it takes a lot longer. Do you want me to enable them?

File: include/pando-lib-galois/graphs/dist_local_csr.hpp (hunk 1 above)
Comment (jeageun) [reply include/pando-lib-galois/graphs/dist_local_csr.hpp:749]: This one is only for one construction. 
Also, how could we know about the  address of target using boolean? 
@divija95

File: include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp (hunk 3 above)
Comment (jeageun) [reply include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp:395]: So, you could branch out from current MDLCSR construction, not the main branch. It just compiled, worked for most of the cases including all test cases. I'm not sure what's the problem....

Comment (jeageun) [reply include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp:395]: Use this as your base, 
https://github.com/utcs-scea/pando-mono/pull/17

Comment (jeageun) [reply include/pando-lib-galois/graphs/mirror_dist_local_csr.hpp:395]: Use https://github.com/utcs-scea/pando-mono/pull/29 instead.

**Pull request merged**